File: bindal/bindal/custom/js/gate_entry.js
frappe.ui.form.on("Gate Entry", {
	refresh: function (frm) {
        if(frm.doc.docstatus){
            // frm.add_custom_button(__('Print Barcode'), function () {
            //   var d = new frappe.ui.Dialog({
            //     title: __('Barcode Details'),
            //     fields: [
            //       {
            //         "label" : "Item",
            //         "fieldname": "item",
            //         "fieldtype": "Link",
            //         "reqd": 1,
            //         "options": "Item",
            //         "get_query":function(r){
            //           return {
            //             query: "bindal.bindal.custom.py.gate_entry.get_items",
            //             filters:{
            //               name:frm.doc.name
            //             }
            //           };
            //         }
            //       },
            //       {
            //         "label" : "Type",
            //         "fieldname": "type",
            //         "fieldtype": "Select",
            //         "options":"Series of Barcode\nSingle Barcode",
            //         "reqd": 1,

            //       },
            //       {
            //         "label" : "Barcode",
            //         "fieldname": "barcode",
            //         "fieldtype": "Link",
            //         "options":"Barcode Label",
            //         "depends_on": 'eval:doc.type == "Single Barcode" ',
            //         "get_query":function(r){
            //           let val = d.get_values()
            //           return{
            //             filters:{
            //               'reference_document':frm.doc.name,
            //               'item': val.item
            //             }
            //           }
            //         }
            //       }
            //     ],
            //     primary_action: function() {
            //       var data = d.get_values();
            //       frappe.call({
            //         method:"bindal.bindal.custom.py.gate_entry.update_barcode",
            //         args:{
            //             doc:frm.doc.name,
            //             data:data
            //         },
            //         callback:function(res){
            //           frm.trigger("load_print_page")
            //         }
            //       })
                
            //     },
            //     primary_action_label: __('Print')
            //   });
            //   d.show();
                
            // });
                
        }
    },
    load_print_page(frm) {
        const print_format = "Raw Material Barcode"
        const letter_head = 0;
        const url =
          frappe.urllib.get_base_url() +
          "/printview?doctype=Gate Entry&name=" +
          frm.doc.name +
          "&trigger_print=1" +
          "&format=" +
          print_format +
          "&no_letterhead=" +
          letter_head;
        const printWindow = window.open(url, "Print");
        printWindow.addEventListener(
          "load",
          function () {
            printWindow.print();
          },
          true
        );
      },
      naming_series(frm) {
        if(frm.doc.naming_series == 'HO-GE26-.####') {
          frm.set_value('branch', 'Head Office')
        }
        else if(frm.doc.naming_series == 'U1-GE26-.####') {
          frm.set_value('branch', 'Unit-1')
        }
        else if(frm.doc.naming_series == 'U2-GE26-.####') {
          frm.set_value('branch', 'Unit-2')
        }
        else if(frm.doc.naming_series == 'U3-GE26-.####') {
          frm.set_value('branch', 'Unit-3')
        }
        else if(frm.doc.naming_series == 'U4-GE26-.####') {
          frm.set_value('branch', 'Unit-4')
        }
      }
})

frappe.ui.form.on("PO Item", {
  custom_total_bundle:function(frm,cdt,cdn){
    let row = locals[cdt][cdn]
    frappe.model.set_value(row.doctype,row.name,"custom_pcs_per_bundle",row.received_qty / row.custom_total_bundle)
  },
  // item_code:function(frm,cdt,cdn){
  //   let row = locals[cdt][cdn]
  //   frappe.db.get_value("Item", row.item_code, "last_updated_series", (r) => {
  //     frappe.model.set_value(cdt, cdn, "last_updated_series", r.last_updated_series);
  //   });
  // },
  // last_updated_series:function(frm,cdt,cdn){
  //   let row = locals[cdt][cdn]
  //   frappe.model.set_value(cdt, cdn, "current_series", row.last_updated_series+1);
  // },
})

File: bindal/bindal/custom/py/gate_entry.py
import frappe
import json
from frappe.model.naming import parse_naming_series
from io import BytesIO
# from barcode import Code128
# from barcode.writer import ImageWriter
import base64


@frappe.whitelist()
def get_items(doctype, txt, searchfield, start, page_len, filters, as_dict=False):    
    return frappe.db.sql(f"""select `tabPO Item`.item_code,`tabPO Item`.item_name from `tabPO Item` where `tabPO Item`.parent = '{filters.get('name')}' and (`tabPO Item`.item_code like '%{txt}%' or `tabPO Item`.item_name like '%{txt}%' )""")


@frappe.whitelist()
def get_barcode(doc):
    doc = frappe.get_doc("Gate Entry",doc)

    for it in doc.item:
        item_doc = frappe.get_doc("Item",it.item_code)
        starting_count = frappe.get_value("Item",it.item_code,'last_updated_series')
        if it.custom_barcode_with_bundle:
            qty = it.custom_total_bundle
        else:
            qty = it.received_qty
        for count in range(starting_count+1,int(qty)+starting_count+1,1):
            barcode = frappe.new_doc('Barcode Label')
            if "U1" in doc.naming_series.replace("-",''):
                unit =  "U1"
            elif "U2" in doc.naming_series.replace("-",''):
                unit = "U2"
            elif "U3" in doc.naming_series.replace("-",''):
                unit = "U3"
            else:
                unit = "WH"
            if it.custom_barcode_with_bundle:
                barcode.barcode = it.item_code +'-'+unit+'-'+ str((doc.name.split("-")[-1]))+'-'+'{:03d}'.format(count) + '-'+str(round(it.custom_pcs_per_bundle))+"-"+it.uom
            else:
                barcode.barcode = it.item_code +'-'+unit+'-'+ str((doc.name.split("-")[-1]))+'-'+'{:03d}'.format(count)

            barcode.reference_doctype = "Gate Entry"
            barcode.reference_document = doc.name
            barcode.item = it.item_code
            barcode.save()
            
            if count == int(qty)+starting_count:
                frappe.set_value("Item",it.item_code,'last_updated_series',count)

    return 1

def on_submit(doc,event):
    get_barcode(doc.name)

@frappe.whitelist()
def update_barcode(doc,data):
    data = json.loads(data)
    frappe.db.set_value("Gate Entry",doc,'barcode_type',data.get('type'))
    frappe.db.set_value("Gate Entry",doc,'barcode',data.get('barcode'))
    frappe.db.set_value("Gate Entry",doc,'barcode_item',data.get('item'))

    # series = frappe.db.get_single_value("Stock Settings","raw_material_series")
    # doc = frappe.get_doc("Gate Entry",doc)

    # for it in doc.item:
    #     if series and not it.barcode:
    #         import re

    #         # Define a function to replace placeholders with their values
    #         def replace(match):
    #             placeholder = match.group(0)  # Get the matched placeholder including curly braces
    #             key = placeholder[2:-2]       # Remove curly braces to get the key (e.g., item.item_code)
    #             # Look up the value in the dictionary, or use the placeholder itself if not found
    #             return it.get(key.replace('item.',''))

    #         # Define a regular expression pattern to match text within curly braces
    #         pattern = r'\.{item.*?\}.'

    #         # Replace all occurrences of the pattern with the values from the dictionary
    #         output_string = re.sub(pattern, replace, series)

    #         html_code = parse_naming_series(output_string,doc=doc)
    #         item_doc =  frappe.get_doc("Item",it.item_code)
    #         item_doc.append("barcodes",{'barcode':html_code})
    #         item_doc.save()
    #         frappe.set_value("PO Item",it.name,'barcode',html_code)

    #         frappe.set_value("Item",it.item_code,'last_updated_series',it.current_series)
            
    #     elif series and it.barcode:
    #         pass
    #     else:
    #         frappe.throw("Series not found for Finished Goods in Stock Settings")
    # return 1